File: 38-count-and-say/count-and-say.js
/**
 * @param {number} n
 * @return {string}
 */
var countAndSay = function (n) {
    if (n === 1) {
        return "1";
    }

    return rle(countAndSay(n - 1));
};

function rle(s) {
    const n = s.length;
    let ans = "";
    let count = 1;

    for (let i = 1; i < n; i++) {
        if (s[i] === s[i - 1]) {
            count++;
        } else {
            ans += count + s[i - 1];
            count = 1;
        }
    }

    ans += count + s[s.length - 1];

    return ans;
}

File: 38-count-and-say/count-and-say.py
class Solution:
    def countAndSay(self, n: int) -> str:
        res = "1"
        for _ in range(n - 1):
            res = self.describe(res)
        return res

    def describe(self, s: str) -> str:
        ans = []
        count = 1

        for i in range(1, len(s)):
            if s[i] == s[i - 1]:
                count += 1
            else:
                ans.append(f"{count}{s[i - 1]}")
                count = 1

        ans.append(f"{count}{s[-1]}")

        return "".join(ans)
